Looks like the test failed:

```
[13:41:06]	[Step 2/2] I170619 13:41:05.721190 1 rand.go:76 Random seed: -5507017542690399957
[13:41:06]	[Step 2/2] I170619 13:41:05.724667 9 gossip/gossip.go:297 [n1] NodeDescriptor set to node_id:1 address:<network_field:"" address_field:"" > attrs:<> locality:<>
[13:41:06]	[Step 2/2] I170619 13:41:05.726493 9 storage/store.go:1269 [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
[13:41:06]	[Step 2/2] I170619 13:41:05.729575 9 storage/replica_command.go:2663 [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
[13:41:06]	[Step 2/2] E170619 13:41:05.732265 9 storage/replica.go:1581 on-disk and in-memory state diverged:
[13:41:06]	[Step 2/2] [RaftAppliedIndex: 11 != 10]
[13:41:06]	[Step 2/2] E170619 13:41:05.732315 9 util/log/crash_reporting.go:82 a panic has occurred!
[13:41:06]	[Step 2/2] --- FAIL: TestStoreSplitGCThreshold (0.01s)
[13:41:06]	[Step 2/2] panic: (log.Safe) (0x1986d20,0xc4201af650) [recovered]
[13:41:06]	[Step 2/2] panic: (log.Safe) (0x1986d20,0xc4201af650) [recovered]
[13:41:06]	[Step 2/2] panic: (log.Safe) (0x1986d20,0xc4201af650) [recovered]
[13:41:06]	[Step 2/2] panic: (log.Safe) (0x1986d20,0xc4201af650) [recovered]
[13:41:06]	[Step 2/2] panic: (log.Safe) (0x1986d20,0xc4201af650)
[13:41:06]	[Step 2/2]
[13:41:06]	[Step 2/2] goroutine 9 [running]:
[13:41:06]	[Step 2/2] testing.tRunner.func1(0xc4203bcd00)
[13:41:06]	[Step 2/2] /usr/local/go/src/testing/testing.go:622 +0x29d
[13:41:06]	[Step 2/2] panic(0x1986d20, 0xc4201af650)
[13:41:06]	[Step 2/2] /usr/local/go/src/runtime/panic.go:489 +0x2cf
[13:41:06]	[Step 2/2] github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
[13:41:06]	[Step 2/2] /go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:69 +0x4b3
[13:41:06]	[Step 2/2] panic(0x1986d20, 0xc4201af650)
[13:41:06]	[Step 2/2] /usr/local/go/src/runtime/panic.go:489 +0x2cf
[13:41:06]	[Step 2/2] github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420362140, 0x7f6758310118, 0xc420016e30)
[13:41:06]	[Step 2/2] /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:200 +0xb1
[13:41:06]	[Step 2/2] panic(0x1986d20, 0xc4201af650)
[13:41:06]	[Step 2/2] /usr/local/go/src/runtime/panic.go:489 +0x2cf
[13:41:06]	[Step 2/2] github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc420362140, 0x7f6758310118, 0xc420016e30)
[13:41:06]	[Step 2/2] /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:447 +0x204
[13:41:06]	[Step 2/2] panic(0x1986d20, 0xc4201af650)
[13:41:06]	[Step 2/2] /usr/local/go/src/runtime/panic.go:489 +0x2cf
[13:41:06]	[Step 2/2] github.com/cockroachdb/cockroach/pkg/storage.(*Replica).assertStateLocked(0xc42053ae00, 0x7f6758310118, 0xc420016e30, 0x27b10c0, 0xc420488160)
[13:41:06]	[Step 2/2] /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1583 +0x5fa
[13:41:06]	[Step 2/2] github.com/cockroachdb/cockroach/pkg/storage.(*Replica).AssertState(0xc42053ae00, 0x7f6758310118, 0xc420016e30, 0x27b10c0, 0xc420488160)
[13:41:06]	[Step 2/2] /go/src/github.com/cockroachdb/cockroach/pkg/storage/helpers_test.go:255 +0xa0
[13:41:06]	[Step 2/2] github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreSplitGCThreshold(0xc4203bcd00)
[13:41:06]	[Step 2/2] /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_split_test.go:1416 +0xe12
[13:41:06]	[Step 2/2] testing.tRunner(0xc4203bcd00, 0x1bdad20)
[13:41:06]	[Step 2/2] /usr/local/go/src/testing/testing.go:657 +0x96
[13:41:06]	[Step 2/2] created by testing.(*T).Run
[13:41:06]	[Step 2/2] /usr/local/go/src/testing/testing.go:697 +0x2ca
```

--- a/pkg/storage/client_split_test.go
+++ b/pkg/storage/client_split_test.go
@@ -1357,6 +1357,65 @@ func TestStoreSplitTimestampCacheDifferentLeaseHolder(t *testing.T) {
 	}
 }
 
+func TestStoreSplitGCThreshold(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	storeCfg := storage.TestStoreConfig(nil)
+	storeCfg.TestingKnobs.DisableSplitQueue = true
+	stopper := stop.NewStopper()
+	defer stopper.Stop(context.TODO())
+	store := createTestStoreWithConfig(t, stopper, storeCfg)
+
+	leftKey := roachpb.Key("a")
+	splitKey := roachpb.Key("b")
+	rightKey := roachpb.Key("c")
+	content := []byte("test")
+
+	pArgs := putArgs(leftKey, content)
+	if _, pErr := client.SendWrapped(context.Background(), rg1(store), pArgs); pErr != nil {
+		t.Fatal(pErr)
+	}
+	pArgs = putArgs(rightKey, content)
+	if _, pErr := client.SendWrapped(context.Background(), rg1(store), pArgs); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	specifiedGCThreshold := hlc.Timestamp{
+		WallTime: 2E9,
+	}
+	specifiedTxnSpanGCThreshold := hlc.Timestamp{
+		WallTime: 3E9,
+	}
+	gcArgs := &roachpb.GCRequest{
+		Span: roachpb.Span{
+			Key:    roachpb.Key(leftKey),
+			EndKey: roachpb.Key(rightKey),
+		},
+		Threshold:          specifiedGCThreshold,
+		TxnSpanGCThreshold: specifiedTxnSpanGCThreshold,
+	}
+	if _, pErr := client.SendWrapped(context.Background(), rg1(store), gcArgs); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	args := adminSplitArgs(splitKey, splitKey)
+	if _, pErr := client.SendWrapped(context.Background(), rg1(store), args); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	repl := store.LookupReplica(roachpb.RKey(splitKey), nil)
+	gcThreshold := repl.GetGCThreshold()
+	txnSpanGCThreshold := repl.GetTxnSpanGCThreshold()
+
+	if !reflect.DeepEqual(gcThreshold, specifiedGCThreshold) {
+		t.Fatalf("expected RHS's GCThreshold is equal to %v, but got %v", specifiedGCThreshold, gcThreshold)
+	}
+	if !reflect.DeepEqual(txnSpanGCThreshold, specifiedTxnSpanGCThreshold) {
+		t.Fatalf("expected RHS's TxnSpanGCThreshold is equal to %v, but got %v", specifiedTxnSpanGCThreshold, txnSpanGCThreshold)
+	}
+
+	repl.AssertState(context.TODO(), store.Engine())
+}
+
 // TestStoreRangeSplitRaceUninitializedRHS reproduces #7600 (before it was
 // fixed). While splits are happening, we simulate incoming messages for the
 // right-hand side to trigger a race between the creation of the proper replica

--- a/pkg/storage/helpers_test.go
+++ b/pkg/storage/helpers_test.go
@@ -32,6 +32,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/internal/client"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/storage/engine"
 	"github.com/cockroachdb/cockroach/pkg/storage/engine/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/storage/storagebase"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -230,6 +231,30 @@ func (r *Replica) ReplicaIDLocked() roachpb.ReplicaID {
 
 func (r *Replica) DescLocked() *roachpb.RangeDescriptor {
 	return r.mu.state.Desc
+}
+
+// GetGCThreshold returns the range's GCThreshold, acquiring a replica lock in
+// the process.
+func (r *Replica) GetGCThreshold() hlc.Timestamp {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return r.mu.state.GCThreshold
+}
+
+// GetTxnSpanGCThreshold returns the range's TxnSpanGCThreshold, acquiring a replica lock in
+// the process.
+func (r *Replica) GetTxnSpanGCThreshold() hlc.Timestamp {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return r.mu.state.TxnSpanGCThreshold
+}
+
+func (r *Replica) AssertState(ctx context.Context, reader engine.Reader) {
+	r.raftMu.Lock()
+	defer r.raftMu.Unlock()
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.assertStateLocked(ctx, reader)
 }
 
 func (r *Replica) RaftLock() {

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -3073,8 +3073,24 @@ func splitTrigger(
 		rightLease := leftLease
 		rightLease.Replica = replica
 
+		gcThreshold, err := makeReplicaStateLoader(rec.RangeID()).loadGCThreshold(ctx, rec.Engine())
+		if err != nil {
+			return enginepb.MVCCStats{}, EvalResult{}, errors.Wrap(err, "unable to load GCThreshold")
+		}
+		if (gcThreshold == hlc.Timestamp{}) {
+			log.VEventf(ctx, 1, "LHS's GCThreshold of split is not set")
+		}
+
+		txnSpanGCThreshold, err := makeReplicaStateLoader(rec.RangeID()).loadTxnSpanGCThreshold(ctx, rec.Engine())
+		if err != nil {
+			return enginepb.MVCCStats{}, EvalResult{}, errors.Wrap(err, "unable to load TxnSpanGCThreshold")
+		}
+		if (txnSpanGCThreshold == hlc.Timestamp{}) {
+			log.VEventf(ctx, 1, "LHS's TxnSpanGCThreshold of split is not set")
+		}
+
 		rightMS, err = writeInitialState(
-			ctx, batch, rightMS, split.RightDesc, oldHS, rightLease,
+			ctx, batch, rightMS, split.RightDesc, oldHS, rightLease, gcThreshold, txnSpanGCThreshold,
 		)
 		if err != nil {
 			return enginepb.MVCCStats{}, EvalResult{}, errors.Wrap(err, "unable to write initial state")

--- a/pkg/storage/replica_state.go
+++ b/pkg/storage/replica_state.go
@@ -509,6 +509,8 @@ func writeInitialState(
 	desc roachpb.RangeDescriptor,
 	oldHS raftpb.HardState,
 	lease roachpb.Lease,
+	gcThreshold hlc.Timestamp,
+	txnSpanGCThreshold hlc.Timestamp,
 ) (enginepb.MVCCStats, error) {
 	rsl := makeReplicaStateLoader(desc.RangeID)
 
@@ -523,11 +525,25 @@ func writeInitialState(
 	}
 	s.Stats = ms
 	s.Lease = &lease
+	s.GCThreshold = gcThreshold
+	s.TxnSpanGCThreshold = txnSpanGCThreshold
 
 	if existingLease, err := rsl.loadLease(ctx, eng); err != nil {
 		return enginepb.MVCCStats{}, errors.Wrap(err, "error reading lease")
 	} else if (existingLease != roachpb.Lease{}) {
 		log.Fatalf(ctx, "expected trivial lease, but found %+v", existingLease)
+	}
+
+	if existingGCThreshold, err := rsl.loadGCThreshold(ctx, eng); err != nil {
+		return enginepb.MVCCStats{}, errors.Wrap(err, "error reading GCThreshold")
+	} else if (existingGCThreshold != hlc.Timestamp{}) {
+		log.Fatalf(ctx, "expected trivial GChreshold, but found %+v", existingGCThreshold)
+	}
+
+	if existingTxnSpanGCThreshold, err := rsl.loadTxnSpanGCThreshold(ctx, eng); err != nil {
+		return enginepb.MVCCStats{}, errors.Wrap(err, "error reading TxnSpanGCThreshold")
+	} else if (existingTxnSpanGCThreshold != hlc.Timestamp{}) {
+		log.Fatalf(ctx, "expected trivial TxnSpanGCThreshold, but found %+v", existingTxnSpanGCThreshold)
 	}
 
 	newMS, err := rsl.save(ctx, eng, s)

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -201,6 +201,8 @@ func (tc *testContext) StartWithStoreConfig(t testing.TB, stopper *stop.Stopper,
 				*testDesc,
 				raftpb.HardState{},
 				roachpb.Lease{},
+				hlc.Timestamp{},
+				hlc.Timestamp{},
 			); err != nil {
 				t.Fatal(err)
 			}

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -1759,7 +1759,7 @@ func (s *Store) BootstrapRange(initialValues []roachpb.KeyValue) error {
 		return err
 	}
 
-	updatedMS, err := writeInitialState(ctx, batch, *ms, *desc, raftpb.HardState{}, roachpb.Lease{})
+	updatedMS, err := writeInitialState(ctx, batch, *ms, *desc, raftpb.HardState{}, roachpb.Lease{}, hlc.Timestamp{}, hlc.Timestamp{})
 	if err != nil {
 		return err
 	}

--- a/pkg/storage/store_test.go
+++ b/pkg/storage/store_test.go
@@ -1006,7 +1006,7 @@ func splitTestRange(store *Store, key, splitKey roachpb.RKey, t *testing.T) *Rep
 	// Minimal amount of work to keep this deprecated machinery working: Write
 	// some required Raft keys.
 	if _, err := writeInitialState(
-		context.Background(), store.engine, enginepb.MVCCStats{}, *desc, raftpb.HardState{}, roachpb.Lease{},
+		context.Background(), store.engine, enginepb.MVCCStats{}, *desc, raftpb.HardState{}, roachpb.Lease{}, hlc.Timestamp{}, hlc.Timestamp{},
 	); err != nil {
 		t.Fatal(err)
 	}
@@ -2261,7 +2261,7 @@ func TestStoreRemovePlaceholderOnRaftIgnored(t *testing.T) {
 	}
 
 	if _, err := writeInitialState(
-		ctx, s.Engine(), enginepb.MVCCStats{}, *repl1.Desc(), raftpb.HardState{}, roachpb.Lease{},
+		ctx, s.Engine(), enginepb.MVCCStats{}, *repl1.Desc(), raftpb.HardState{}, roachpb.Lease{}, hlc.Timestamp{}, hlc.Timestamp{},
 	); err != nil {
 		t.Fatal(err)
 	}

